### PR TITLE
chore: deps husky and commitlint

### PR DESCRIPTION
**Pull request by balaji-sivasakthi**
### Related issue:-
Issue #16 
### Changes Made:-
- commitlint with husky config
- lint-stage config
- readme and contributor markdown updated
### Additional Context:- 

| Type      | Heading  | Rule                                      | Description                                  |
| --------- | -------- | ----------------------------------------- | -------------------------------------------- |
| ci        | CI       | Continuous Integration                    | Changes related to continuous integration.   |
| chore     | Chore    | Maintenance tasks                         | Other changes that don't affect production.  |
| docs      | Docs     | Documentation                             | Changes related to documentation.            |
| feat      | Feature  | New Feature                               | New feature implementations or additions.    |
| fix       | Fix      | Bug Fixes                                 | Bug fixes or corrections.                    |
| perf      | Perf     | Performance Improvements                  | Performance-related improvements.            |
| refactor  | Refactor | Code Refactoring                          | Code changes that don't fix bugs or add features, but improve the code structure. |
| revert    | Revert   | Revert Previous Commits                   | Reverting previous commits.                   |
| style     | Style    | Code Formatting or Style                  | Changes related to code formatting or style. |
| assets    | Assets   | Add or Update Assets (e.g., images, files) | Changes related to adding or updating assets, such as images or other files. |

### CODE REVIEW

**Comment by wajeshubham** [comment]
@balaji-sivasakthi I appreciate your effort in keeping the project and commit messages tidy. 

However, upon testing the implementation, I discovered that the `git commit` command does not fail when an ugly commit message is provided, which ideally should happen to enforce clean commit messages. 

Also, is it expected that i run `yarn pre-commit` to lint my staged files every time before committing? 

Getting the following warning upon committing.

```
hint: The '.husky/pre-commit' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
hint: The '.husky/commit-msg' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
```
Please let me know your thoughts on this matter, and if you agree with these points. I would love to learn more about this.

**Comment by balaji-sivasakthi** [comment]
@wajeshubham Thank you!  Husky will handle the pre-commit on its own, so there's no need to execute 'yarn pre-commit' before every commit. However, there might be some edge cases where we can use `yarn pre-commit` with specific options, such as `yarn pre-commit --allow-empty` to style the staged files without needing a commit message.

Sorry for the trouble, It was an egde case in mac/Linux
```
hint: The '.husky/pre-commit' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
hint: The '.husky/commit-msg' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
```
To prevent this issue, we can allow permission to the `.husky` directory by executing the following command:

```bash
chmod ug+x .husky/*
```

This command grants execute (`x`) permission to the user (`u`) and the group (`g`) for all files (`*`) inside the `.husky` directory. It ensures that the necessary scripts in the `.husky` directory can be executed properly during the pre-commit process.

```json
{
"prepare": "husky install | chmod ug+x .husky/*"
}
```

**Comment by wajeshubham** [comment]
So, how about adding `yarn prepare` or `npm run prepare` in `./husky/commit-msg`? So, that we don't need to manually install the husky hooks.
```sh
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"
yarn prepare # prepare husky hooks
npx --no-install commitlint --edit
```
Also, `"prepare": "husky install | chmod ug+x .husky/*"` This script works for macOS. However, just to be sure this script should not break in the windows OS.

NOTE: As this PR has changes that are involved in the contribution flow we also need to update the Contribution guide according to this new convention of commit messages.

**Comment by balaji-sivasakthi** [comment]
```json 
"prepare": "husky install | chmod ug+x .husky/*"
```
Yes, this script would break on Windows. However, it can be sorted out by using cross-spawn for executing the prepare script.

```js
import { spawnSync } from 'child_process';
import os from 'os';

// Function to run commands depending on the OS
const runCommand = (command, args) => {
  const isWindows = os.platform() === 'win32';
  const shell = isWindows ? true : false;
  const commandArgs = isWindows ? ['/C', command, ...args] : [...args];

  const result = spawnSync(isWindows ? 'cmd' : command, commandArgs, {
    stdio: 'inherit',
    shell,
  });

  return result.status;
};

// Function to prepare Husky
const prepareHusky = () => {
  // Run husky install
  const huskyInstallStatus = runCommand('npx', ['husky', 'install']);

  // Set permissions for husky on macOS/Linux
  if (os.platform() !== 'win32') {
    const setPermissionsStatus = runCommand('chmod', ['+x', '.husky/*']);
    if (setPermissionsStatus !== 0) {
      console.error('Setting permissions failed');
      process.exit(1);
    }
  }

  // Exit with appropriate status code based on husky install status
  if (huskyInstallStatus === 0) {
    console.log('Husky preparation executed successfully!');
    process.exit(0);
  } else {
    console.error('Husky installation failed');
    process.exit(1);
  }
};

prepareHusky()
```

> NOTE: As this PR has changes that are involved in the contribution flow we also need to update the Contribution guide according to this new convention of commit messages.

Ok, Sure I will add all necessary docs.

**Comment by wajeshubham** [comment]
@balaji-sivasakthi script is failing in the macOS on running `yarn prepare`

```bash
# Error:
husky - Git hooks installed
chmod: .husky/*: No such file or directory # I have .husky folder. Testing your branch only
Setting permissions failed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
Also, seems like an extra boilerplate to just lint the commit, which no doubt is a good practice. However, it would be great if we consider and test all the edge cases before merging.

Thank you! 😀

**Comment by balaji-sivasakthi** [comment]
@wajeshubham 

```
# Error:
husky - Git hooks installed
chmod: .husky/*: No such file or directory # I have .husky folder. Testing your branch only
Setting permissions failed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
I've tested it in **GitHub Codespace**, and it works well now.

**Comment by balaji-sivasakthi** [comment]
@wajeshubham I'm also updated README.md and CONTRIBUTOR.md for conventional commit message format. Please take a moment to look at this.

**Comment by jwala-anirudh** [comment]
@balaji-sivasakthi thanks for the PR, I will review it thoroughly and provide the approval

**Comment by balaji-sivasakthi** [comment]
@wajeshubham Is this ok?

**Comment by wajeshubham** [comment]
Yes @balaji-sivasakthi we are reviewing it.

**Review by jwala-anirudh — CHANGES_REQUESTED** [review]
I have added some comments after reviewing the changes, please go through them

File: .husky/pre-commit [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npx lint-staged --allow-empty
```
[the remark this answers is not in the document]
Comment (balaji-sivasakthi) [reply .husky/pre-commit:4]: It was correctly configured in `.lintstagedrc`, also I installed `lint-staged` with devDependencies
```json
  "devDependencies": {
    "@commitlint/cli": "^17.6.6",
    "@commitlint/config-conventional": "^17.6.5",
    "husky": "^8.0.3",
    "lint-staged": "^13.2.2",
    "nodemon": "^2.0.22",
    "prettier": "^3.0.0"
  }
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -11,10 +11,37 @@ To contribute to FreeAPI, please follow these guidelines:
 1. Fork the repository on GitHub.
 2. Clone your forked repository to your local machine.
 3. Create a new branch for your feature or bug fix: `git checkout -b feat/your-feature-name` or `git checkout -b fix/your-bug-fix-name`.
-4. Make your changes and commit them with descriptive messages: `git commit -am 'Add your commit message'`.
+4. Make your changes and commit them with descriptive messages: `git commit -am 'add your commit message'`. ([Commit format guidelines](#commit-message-format))
```
[the remark this answers is not in the document]
Comment (balaji-sivasakthi) [reply CONTRIBUTING.md]: This has been resolved. Please check now.

File: prepare.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+// Function to prepare Husky
+const prepareHusky = () => {
+  // Run husky install
+  const huskyInstallStatus = runCommand("npx", ["husky", "install"]);
+
+  // Check if .husky directory exists, if not, create it
+  if (!fs.existsSync(".husky")) {
+    fs.mkdirSync(".husky");
```
[the remark this answers is not in the document]
Comment (balaji-sivasakthi) [reply prepare.js]: I added `fs.mkdir()` asynchronously with proper error handling, which would also be a good alternative to `mkdirSync`, as it won't block the event loop and can handle errors more gracefully.

**Comment by balaji-sivasakthi** [comment]
@jwala-anirudh Please check the recent changes

**Review by jwala-anirudh — APPROVED** [review]
Approved.

**Review by wajeshubham — APPROVED** [review]

**Review by wajeshubham — APPROVED** [review]
Looks good!
